### DISCUSSION TITLE
An update on NGINX TCP termination and questions about scaling and high availability features for EG

### DISCUSSION THREAD

**Discussion by averageflamethrowerguy** [jupyter/enterprise_gateway#949] [Q&A] [answered]
Hey Kevin,
I know it's been a ridiculously long time since I asked you about the Nginx TCP issue on EKS. I figured you may still be interested in the documentation of my solution.

This issue will come up if you're deploying the Kubernetes version of Jupyter Enterprise Gateway to AWS EKS. Standard REST operations will succeed, but if you try to create a websocket via wss://your_gateway_url/gateway/api/kernels/${kernelId}/channels, the operation will fail. [This issue]( https://github.com/kubernetes/ingress-nginx/issues/3746) is very similar.

The TLDR of the issue was that the standard Nginx config sets the AWS load balancer to downgrade TCP requests to http within a cluster. However, EKS requires TCP requests in order to upgrade to websockets. (I don't claim to fully understand this, but changing this configuration does work). I attached a document sheet, which may be useful. Feel free to use it if it is.

[EKS_Nginx_websockets.txt](https://github.com/jupyter/enterprise_gateway/files/6214833/EKS_Nginx_websockets.txt)

Onto my question about scaling:
Recently I scaled up my EG cluster to support up to 300 users, and I started noticing some performance issues. GET requests and websocket creation on the cluster are taking ~5 seconds and ~15 seconds, respectively, when they were both nearly instant at smaller cluster sizes. My intuition was that this was an issue with the hub pod, but the problem persisted even when I dramatically increased the resources assigned to that pod. I'm wondering what scaling difficulties you've run into previously, and what number of pods EG is designed to support.
One potential solution that interested me is using multiple replicas of EG and routing traffic between them based on pod id within the REST request. I'm curious about whether you or anyone else on the team is working on something like this.

It turns out that EG has fit the project that I'm working on really well. You can check it out here if you're interested: https://qbook.qbraid.com. Thank you so much for building out such an awesome tool!

**Comment by kevin-bates** [comment]
Hi Elliot - it's good to hear from you.  Thank you for the detailed information regarding the AWS LB and EKS.  I'm sure this will be useful to others.

Regarding scaling, we have not spent time directly with testing the various thresholds but have heard of installations pushing boundaries above 300 - although I'm not certain we have the details.  It would be good to understand the GET and WebSocket issues a bit better.  What version of EG and Notebook are you running against?

It's tough to have a discussion about HA without also involving disaster recovery (DR).  Because we set [`sessionAffinity` to `ClientIP`](https://github.com/jupyter/enterprise_gateway/blob/master/etc/kubernetes/helm/enterprise-gateway/templates/service.yaml#L22), which should route clients to the same EG instance, **you should be able to increase replicas today**.  This would be a pure HA play.  (Have you tried increasing the replicas?)

We also have an experimental _kernel session pesistence_ option (`--KernelSessionManager.enable_persistence=True`) which will maintain active kernel session information in a file (presumably on a shared file system - with the idea a NOSQL DB be used for this) such that if one hosting EG goes down, another EG can be started at which time, the remote kernel sessions registered in the file will be reconnected from the newly started EG.  This would be a pure DR play. (active/passive)

What still needs to happen to have both HA/DR is we need to intercept "kernel-not-found" exceptions in the kernel handler and go through the motions of checking the persisted kernel sessions, such that if the desired kernel id exists, then it should be loaded on _that_ EG server and reconnected.  This would allow for a more "active/active" behavior.  However, in this mode, we would likely need to disable the previous DR functionality - where _starting_ EG causes _all_ of the persisted sessions to be reconnected.  In this active/active approach, we would let the user trigger the reconnection by issuing requests against a remote kernel that the original EG instance cannot satisfy.  Sine the new EG doesn't know about that kernel, it would reconnect on the fly.

Other details that would need to be worked out on this are how culling would work.  Since the kernel's last activity (which is used to determine the kernel's idle period) is maintained in memory, then when a new EG takes over management, the culling period would reset.  That isn't so bad since the kernel would be culled eventually.  However, if the user never triggers the kernel's "hydration" on the new EG instance (i.e., the kernel was destined to be culled), then, it may never be culled and could result in orphaned kernel pods/processes.

One approach could be to also record the kernel's last activity in its persisted information, then when the new EG hydrates the kernel, the last activity maintained in memory could be used to seed the culling monitor.  We could also have each EG instance look for culling candidates to address the cases where the user's action didn't hydrate the kernel.  But again, that solution would require we update the kernel's persistent state on each activity and I'm not sure that's something we should do from a performance perspective.

We've talked about this kind of stuff on https://github.com/jupyter/enterprise_gateway/issues/562

I think adding the _kernel-not-found_ code to check the persisted information would be straightforward and useful.  Perhaps we could introduce some kind of DR-style option (active/active or active/passive) so the two flavors could co-exist.  In active/active, EG would not load persisted sessions from the file, but let their "miss" trigger reconnection.  In active/passive, newly started EGs would load persisted sessions.  Since the idea is that there would never be more than one active EG, the "miss/reconnect" code could still exist since it should, in theory, never find a missing kernel id that is in the persisted state.

Sorry for the ramble, but I believe those are the current thoughts on where were are on the HA/DR front.

**Reply by averageflamethrowerguy** [reply]
I'm running the elyra/enterprise-gateway:dev image on the EG, and running a custom kernel built off the jupyter/scipy-notebook:04f7f60d34a6 image, which may be a little old (I just installed some additional Python packages).
I'm provisioning 1.5vCPU and 3Gb RAM for the EG pod and 0.12vCPU and 0.6Gb per kernel. I realize that these are very low resources for the kernel. I have done some testing on larger allocations, which didn't seem to make a difference, but I haven't conclusively demonstrated that the problem isn't resource starvation. I'll probably do more testing tonight.

We're using a simple JavaScript session management system based largely on [the Jupyter EG session manager](https://github.com/jupyter/notebook/tree/master/notebook/gateway) you linked me to a while back. I can send you the implementation if you'd like.

I did try increasing the EG pod replicas (to 5). What I found was that I wasn't consistently querying the same EG pod; i.e, I could start a kernel by doing a POST to ${BASE_URL}/gateway/api/kernels to start a kernel and retrieve the kernel_id, but then I would get a 404 when doing a GET to ${BASE_URL}/api/kernels/${kernel_id}. Further poking around showed that the pod responding to the query was inconsistent (I was cycling between pods). 
My intuition says that the ClientIP in question here should be the IP of the frontend (or backend) client sending this request. Is this right? If so, my hunch is that NGINX isn't correctly passing the header forward, which results in the switching behavior I was seeing.

It doesn't seem like you would need a particularly expensive process to make sure we can cull non-hydrated kernels. Unless I'm missing something, it seems like an EG would only remove a culled kernel from the database when it confirms the culling to be complete (either via a GET request or kubectl command perhaps; I'm not sure how culling works precisely). Regardless, if the EG died while a kernel was culling, the next EG should be able to still use the database reference to determine whether the kernel was successfully culled or not. The only hiccup is you can no longer assume that any kernel in the database necessarily still exists -- but that's not really a safe assumption anyway, I think.

#562 is pretty great stuff (as is [#594](https://github.com/jupyter/enterprise_gateway/issues/594)).
What I gather is the current DR approach is unsuitable for HA because we currently don't have a good way of coupling EG lifecycle with those of kernels, and end up reconnecting all kernels to all restarted EGs. Your proposed active/active solution is interesting here. Is the idea that a newly spawned EG searches for an orphaned kernel only when a query against that kernel occurs (basically, when does the miss event happen)? If so, we'd still need a culling solution that searches against persistent state, in order to clean up old orphaned kernels, but that could be invoked rarely, and therefore shouldn't be that expensive.

**Reply by kevin-bates** [reply]
> I'm running the elyra/enterprise-gateway:dev image on the EG,

Hmm.  This makes it very difficult to determine what code is actually running as the `dev` images can change at any time - beware.

> My intuition says that the ClientIP in question here should be the IP of the frontend (or backend) client sending this request. Is this right?

I believe so.  When I ran with multiple replicas a while back, that's the behavior I saw.  I have extremely limited experience with NGINX, but that sounds like a good guess that it's blocking something.   If you can make a given client have affinity to a given EG, then there should be nothing preventing multiple replicas from working out of the box.

I'm not sure you have the correct understanding of what culling is in the jupyter server context.  All that it is is that the server will periodically (the culling interval) check the kernel's last activity timestamp (which is stored in memory within the MappingKernelManager) against a configured idle timeout value.  If the kernel's last activity occurred beyond the idle timeout value, then that kernel will be terminated (culled).  The check at each culling interval is implemented via a periodic callback.  This is pure jupyter code and has nothing to do with the hosted env (e.g., kubernetes, etc.).

When we speak of a single EG server, there are really no issues.  That single EG server manages all kernels - even it was started due to DR, all kernels are hydrated by _that_ EG instance.  Granted, because the kernel's last activity is not persisted, then when the first EG crashed, it took within the last activity of each kernel, so, when the DR EG hyrdrates the kernels from their persisted state, it essentially resets the last-activity, thereby resetting the kernel's cull time to be a full idle-timeout duration.

Where things get more involved is when there are multiple EG's running we don't (today) know which of the persisted kernel states are currently being managed by an EG instance.  This only becomes a problem when an EG fails.  Yes, if the user attempts to access an orphaned kernel, their session must hit an EG, and EG will trap the 404 and reconnect to the kernel using its persisted state.  But the issue is with the orphaned kernels to which the user will never issue a GET request again.  Because one EG cannot determine if a given persisted kernel is being serviced by another EG, having a "sweep" mechanism isn't sufficient unless we also persist the kernel's last activity timestamp each time the kernel is active.  But even then, you could get into a race condition with the _owning_ EG's culler.

I think some kind of registration that an EG performs may be useful.  That is, when it starts managing a kernel, it records that fact in a shared location.  You'd then also need some form of heartbeat mechanism because another EG cannot necessarily determine if the _owning_ EG is still running. So things get a little messy fairly quickly.

If each EG knew the URIs for the others, then there could be a way to determine if a given kernel id is being managed or not by querying for the set of kernels across the EG instances.  A separate single-instance service that looks out for the orphaned kernels seems like a good approach for that.

I know this response is fairly scattered and I apologize.

**Reply by averageflamethrowerguy** [reply]
This clarifies a lot, yes! And don't apologize -- we're discussing a couple of things simultaneously, and that's in large part my fault for not segmenting threads. You should also let me know if I should read more threads and docs before continuing this discussion. I've only done a limited amount of that, and I'm worried that I'm bringing up non-issues or things you've already solved. 

You're right, I did misunderstand the culling behavior. I had definitely assumed that more information about state was drawn from the kernels themselves. I see why you'd do it this way -- its way more efficient, particularly for single-EG cases, which are probably the majority of deployments.

Having the separate single-instance service might be a pretty good strategy. You still would need some sort of timestamped information -- it could be a periodic operation the EG does to say its still managing a particular kernel (as you were describing). The service would cull kernels on an interval 2 * (eg_update_interval + kernel_timeout) or something. The idea is that the service would have a longer timeout to allow any managing EG instances to do cull behavior without risk of race conditions. If kernels in the store get really stale and exceed the service culling threshold, we should be able to safely cull them without interaction from managing EGs (the only risk is an EG connects to a kernel just as its being culled). This would allow us to clean up orphaned kernels without the service necessarily knowing EG URIs.

We may run into an issue with the current HA scheme, and connecting to the same kernel with multiple clients. If the ClientIP does end up being user-specific, then we can't guarantee users are routed through the same EG. Imagine we have two EGs, one kernel,  and two users. The first user may connect to the kernel by routing through EG 1 and specifying the kernel ID; EG 1 will connect to the kernel if it hasn't already. However, if the second user tries to connect to the kernel via EG 2, and EG 2 isn't blocking based on existing EG 1's connections to kernels (or the persistent store hasn't updated yet), then EG 2 would _also_ connect to the kernel, and we'd run into issues with both EGs managing kernel state. On the other hand, if EG 2 refused to connect to the kernel because the persistent store shows that EG 1 already manages that kernel, then the second user is unable to connect to a kernel that exists. This may be bad if said user was using the kernel through a previous EG that died.

This is perhaps most clearly solved with a router that sends traffic to the EG with the desired kernel. This would end up relying on very up-to-date persistent storage of kernel-EG pairs, and would also need to know EG URIs. (Incidentally, these requirements help with one of the culling schemes you proposed...)

Thanks for the heads-up about the dev image. I'll go onto a hard-coded version.

**Reply by averageflamethrowerguy** [reply]
I do think that the case where EG 2 refuses to connect to an existing kernel only is a problem relatively rarely. and is very dependent on frontend behavior (i.e, the kernel provisioner has to figure out that the websocket connection was broken, and then reconnect to the same kernel). So a valid choice would just be to ignore it and force user 2 to spin up a new kernel.

**Reply by averageflamethrowerguy** [reply]
Just verified slow response times using much higher resources per kernel.
Doing a GET request to get all kernels ~5 secs; to get a particular kernel ~5 secs, to provision a websocket 10-12 secs. This is with kernels with 0.5 vCPU and 1 Gb RAM each, a spec that is fine on our JupyterHub cluster. This is in-line with what I was seeing at lower resource allocations, so I'm pretty sure that the bottleneck isn't pod resources.
A test of the same number of pods (300) across 10 EBs gives good performance, with near-instant HTTP requests and websockets (though I need to solve the Nginx issue before its useful). So the performance problem is something to do with scaling a particular EB, rather than the entire cluster.

**Reply by kevin-bates** [reply]
Hi Elliot, is this something you'd be interested in contributing to?  If so, it might be helpful to set up a Webex meeting where we can discuss things and draft up a more detailed design.  My colleagues and I are in the Pacific timezone and typically find that afternoons work best.

**Reply by averageflamethrowerguy** [reply]
It certainly is!
I would suggest we meet sometime 2-6pm PDT on either Tuesday or Wednesday.

**Reply by kevin-bates** [reply]
Hi Elliot, I've sent you a meeting invitation for Wednesday from 3 - 4 pm (PDT).  

For others that might be interested, please feel free to join as well at: https://ibm.webex.com/meet/kevin.bates

We'll continue the discussion here focusing on these items:
- Replicas and affinity
- Active/Active vs. Active/Passive
- Culling and orphan detection

**Reply by kevin-bates** [reply]
Hi @averageflamethrowerguy - a last-minute change has cropped up for today.  Any chance we could push our meeting to 4:30 (PDT) or tomorrow (Thursday, April 1) at 3:00 pm (PDT)?  Sorry for the inconvenience.

**Reply by kevin-bates** [reply]
UPDATE: Today's meeting has been moved to 4:30 (PDT).

**Comment by averageflamethrowerguy** [comment]
Starting a new subthread here about NGINX stuff and networking within Kubernetes / EG.
As I mentioned in the earlier subthread, I'm running into an issue with scaling multiple EG instances behind an NGINX proxy. The desired behavior is that a particular client is routed to the same EG instance upon subsequent HTTP requests (i.e, I can do a GET to get all active kernels, then a GET to get a particular kernel, and the second GET will always succeed because I'm on the same EG (which manages the same kernels).

Before you get buried in the details, I'd like to ask a few questions:
1. Do you know which header sessionAffinity: ClientIP looks at when trying to determine ClientIP?
2. Are there tools you've used before to analyze network traffic within a Kubernetes cluster? (Basically, I need to figure out what the headers transmitted between the Kubernetes ingress controller and the NodePort service are, and then what the headers downstream of that are).
Failing that, how would I look at the headers of the requests actually hitting EG (this could either be a debug flag, or you could point me in the direction of where your GET handler directly uses an HTTP library, and I'll modify that within an active image for now (rather than doing a rebuild).
3. If you got a multi-EG deployment to work somewhere else, could you tell me how you set that up (i.e, what cloud provider, were you using a LoadBalancer directly or pairing it with an ingress? It's easily possible that I'm running into an issue you already solved or dodged entirely.)
4. Is there any way you know to configure session affinity with a cookie within a Service? 

So from my research, here's a description of the symptoms of the problem and the architectures involved.

The problem:
Using AWS EKS (I haven't tested with any other cloud provider) setting `sessionAffinity: ClientIP` within the selector of the service.yaml file doesn't produce sticky sessions as desired.
Instead, it looks like I'm routed to different EGs in a round-robin style approach. (I'm also routed RR when sessionAffinity: ClientIP is not specified.) 

I started by investigating the NGINX LoadBalancer service/controller. In order to propagate information about the client IP address, my understanding is we need to enable proxy protocol both in the Load Balancer and in the ConfigMap.
```
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "tcp"
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
spec:
  type: LoadBalancer
<additional configuration omitted for brevity>
```

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ingress-nginx-controller
  namespace: ingress-nginx
  labels:
      helm.sh/chart: ingress-nginx-3.10.1
      app.kubernetes.io/name: ingress-nginx
      app.kubernetes.io/instance: ingress-nginx
      app.kubernetes.io/version: 0.41.2
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/component: controller
data:
  http-snippet: |
    server {
      listen 2443;
      return 308 https://$host$request_uri;
    }
  real-ip-header: "proxy_protocol"
  proxy-read-timeout: "3600"
  proxy-send-timeout: "3600"
  set-real-ip-from: "0.0.0.0/0"
  proxy-real-ip-cidr: <omitted>
  use-forwarded-headers: "true"
  force-ssl-redirect: "false"
```

After doing this configuration step, I verified that both the  X-Real-IP and X-Forwarded-For headers were being set. You can do this by increasing NGINX log level using [this guide](https://kubernetes.github.io/ingress-nginx/troubleshooting/).
However, the problem still persists.

The next thing I tried was changing the affinity of the NGINX ingress itself. You can do this by specifying the following in values.yaml:
```
  nginx:
    enabled: true
    path: /gateway/?(.*)
    annotations:
      nginx.ingress.kubernetes.io/affinity: "cookie"
      nginx.ingress.kubernetes.io/session-cookie-expires: "172800"
      nginx.ingress.kubernetes.io/session-cookie-max-age: "172800"
      nginx.ingress.kubernetes.io/session-cookie-name: "route"
      nginx.ingress.kubernetes.io/affinity-mode: "persistent"
```
Unfortunately, as far as I can tell, NGINX only supports cookie affinity and Kubernetes Services only support ClientIP affinity.
NGINX will set a cookie on the client and use that for routing. My understanding is this routing happens at the Ingress level -- i.e, immediately behind the load balancer service.
Setting this had a very interesting effect. Alone, (i.e, without `sessionAffinity: ClientIP`) I get _random_ successes and failures. For instance, if I'm doing a GET for a particular kernel, I will randomly succeed with probability ~1/$number_of_pods. This is in comparison to round-robin assignment, where I'd always succeed periodically (with 5 pods, I'd succeed the 5th try, 10th try, 15th try ..)
When paired with ClientIP affinity, I get _consistent_ successes and failures within a particular request type. By that, I mean that _all_ GETs to a particular kernel will succeed or fail. This is desired behavior. But if I did a POST to spin up a kernel and then a GET to find the kernel, the GET will generally fail (and fail consistently after that). (basically, it seems like the session is sticky within the same type of request). It's almost like I'm failing to forward the cookie or something when I switch request types...

The last piece of configuration I tried is related to routing within the Service created by service.yaml. My intuition is that setting an affinity at the Service level _should_ be sufficient for sticky sessions. So I suspected that perhaps the headers weren't reaching the service.
Once possibility is actually at the service level itself. The Service declared in service.yaml is of type `NodePort`. NodePort services typically obscure clientIP information because they maintain their own routing tables for networking purposes. Headers downstream refer to the IP in the table.
Supposedly, you can get around this by setting
```
spec:
     externalTrafficPolicy: Local
```
but this ultimately didn't change any of the behavior I was experiencing. This isn't really surprising: because we are setting `sessionAffinity: ClientIP` within the same service, it is reasonable to expect the Service handles stickiness prior to obfuscating IP headers (though I'm by no means sure that this is what's going on).

Here's a diagram and some [light reading](https://oteemo.com/ingress-101-what-is-kubernetes-ingress-why-does-it-exist/) if you are as confused as I was about Ingress vs Load Balancers.
![ingress-controller](https://user-images.githubusercontent.com/58124535/113352382-297dd300-930a-11eb-8b4b-9b958bf02192.png)

**Reply by kevin-bates** [reply]
Hi Elliot - this sounds very frustrating and I really don't have anything useful to provide.  We, as maintainers, really don't run EG in cloud based settings, we tend to configure a cluster of servers to ensure the minimally-viable functionality is working.  

I really don't have the bandwidth to dig deep into this (besides the fact that I have very limited knowledge of this aspect of things).  In looking around just now, I believe I can confirm that NGINX uses "cookie" affinity while k8s uses "clientIP".  I have seen some implications that an NGINX controller can hit pods directly, but I don't understand how that would be conveyed/configured (the idea being we'd not use the EG service - not even sure if that's an option).

I believe @lresende has worked with traefik ingresses in the past and I'm not sure if that's an option for you, nor what kinds of affinity characteristics it might provide.

I'm sure you've exhausted the web searches, but want to share a couple I found in case you find them helpful:

https://stackoverflow.com/questions/59272484/sticky-sessions-on-kubernetes-cluster

Issue 1980 in this repo: https://github.com/kubernetes/ingress-nginx/issues/. (you might find other issues there that might be helpful)

I'm sorry to leave you hanging on this - I'm not knowledgeable enough in network configuration.  I will try to ask around and see if I can locate a resource that might be able to help.

**Reply by averageflamethrowerguy** [reply]
Yeah no worries -- its definitely very arcane stuff!
Mostly the information I posted here is just a reference for anyone else fiddling with this -- I don't really expect you to dig into this or any other TED talk I will inevitably post here. Mostly, if you don't know the answer to a question or where to find it, just let me know rather than searching for it.
I think Traefik is definitely an option -- I ended up steering clear of it a few months ago because it seemed like the less robust ingress offering compared to NGINX. Ultimately, I'm betting the issue is not with the load balancer or ingress (because I saw the correct headers in the log files) but is downstream somewhere around the NodePort service.

Anyway, I'd still really appreciate it if you could give me direction about where the basic network stuff in EB are located. (I think you're using Tornado, for instance?). I'm going to try to get into an EG and log headers from there, which should give me a good sense of whether headers are being passed cleanly through the entire system. If you haven't poked around with this before, no worries.

**Reply by kevin-bates** [reply]
> Anyway, I'd still really appreciate it if you could give me direction about where the basic network stuff in EB are located. (I think you're using Tornado, for instance?). I'm going to try to get into an EG and log headers from there, which should give me a good sense of whether headers are being passed cleanly through the entire system. If you haven't poked around with this before, no worries.

Yes, all Jupyter applications are built on `tornado` (for better or worse).

The kernel-based handlers are defined in [`enterprise_gateway.services.kernels.handlers.py`](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernels/handlers.py) while the kernelspec handlers are defined in [`enterprise_gateway.services.kernelspecs.handlers.py`](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernelspecs/handlers.py).

You're probably most interested in the kernel startup and fetch of a running kernel.  The handler for the startup (POST request) is in the [`MainKernelHandler`](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernels/handlers.py#L31), while the handler for fetching a known kernel is in the [`KernelHandler`](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernels/handlers.py#L130).

I just set a breakpoint in the [kernel**spec** handler](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/services/kernelspecs/handlers.py#L64) (I find that API helpful because it should always return something and doesn't trigger any actions) to determine some log statements you could add in the various methods.  To dump the Cookies and Headers, you should find these statements helpful:
```python
        self.log.debug(f"Cookies: {self.request.cookies}")
        self.log.debug(f"Headers: {self.request.headers}")
```
which, for my local laptop, produced the following kinds of output...
```
[D 2021-04-02 07:22:58.769 EnterpriseGatewayApp] Cookies: Set-Cookie: _xsrf=2|4bbfa8f1|0e0b28ffde10152128998c75a9ad22e9|1617203438
    Set-Cookie: username-localhost-8888="2|1:0|10:1617203438|23:username-localhost-8888|44:NzkwMjFlNNDLSU2MzIzNDA2ZjljYzcxZTc4OWZiYjgzNTI=|dc74fe88hjkIOU81fe7e680d5c4f1bca7b6648c1394f6d94985784172d9c9eab9a96"
[D 2021-04-02 07:22:58.770 EnterpriseGatewayApp] Headers: Host: localhost:8889
    User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:86.0) Gecko/20100101 Firefox/86.0
    Accept: */*
    Accept-Language: en-US,en;q=0.5
    Accept-Encoding: gzip, deflate
```
`self.request` is an [`HTTPServerRequest`](https://www.tornadoweb.org/en/stable/httputil.html#tornado.httputil.HTTPServerRequest) instance, while `self.request.cookies` and `self.request.headers` are [`SimpleCookie`](https://docs.python.org/3/library/http.cookies.html#http.cookies.SimpleCookie) and [`HTTPHeaders`](https://www.tornadoweb.org/en/stable/httputil.html#tornado.httputil.HTTPHeaders) instances respectively.

Thank you again for taking time to dig in and understand this - it will be good that somebody understands. :smile:

**Reply by kevin-bates** [reply]
Hi Elliot, I wanted to let you know that I've reached out to some colleagues and they're still getting up to speed on the problem definition.  However, it occurred to us while talking if we need NGINX to perform any affinity operations at all since it really is only "routing" to the _single_ EG service - which, itself, is what is fronting the replicated deployments (replicas).

So, as you surmised a while ago, this seems to be an exercise of understanding how the k8s service determines the value of the ClientIP for its sessionAffinity attribute.  That said,  do wonder if attempting to configure any affinity in NGINX might be exacerbating the real issue.  Your analysis seems sound though.

I'm copying my colleagues @delapuente and @akchinSTC in case they have any further questions or insights.  

I appreciate everyone's help on this!

**Reply by averageflamethrowerguy** [reply]
Sweet! I appreciate you reaching out to help solve this! 
Yes, messing with NGINX affinity was something I did before I had much of a grasp of the networking components involved. It's probably a red herring, but it changed the symptoms of the problem and was therefore interesting.

**Reply by averageflamethrowerguy** [reply]
Alright, I did some fiddling around with the logs tonight. Unfortunately, I wasn't able to get much information.
Here's what I've been working with (feel free to suggest a different workflow, because I feel like I'm doing this in an inefficient way ...)
I'm manually modifying an EG pod deployed to the EKS cluster in order to try to get logs out of it. I've tried modifying the kernelspec handlers in both /opt/conda/lib/python3.7/site-packages/enterprise_gateway/services/kernelspecs/handlers.py and /opt/conda/lib/python3.7/site-packages/notebook/services/kernelspecs/handlers.py. For both of these, I put the logs just within the get() handlers. (I also put a log [just over one I knew was firing](https://github.com/jupyter/jupyter_client/blob/706f61efafe00f3255e6b42e8d789764a468f114/jupyter_client/kernelspec.py#L181)  in /opt/conda/lib/python3.7/site-packages/jupyter_client/kernelspec.py).

None of these new logs ended up showing up in my debug logs.
Does the build step end up copying any of these files into a different directory that is then active?
I confess to being at a bit of a loss.

**Reply by kevin-bates** [reply]
>  I put the logs just within the get() handlers. (I also put a log just over one I knew was firing in /opt/conda/lib/python3.7/site-packages/jupyter_client/kernelspec.py).

> None of these new logs ended up showing up in my debug logs.

If you're editing the running pod instance of EG, how did you restart it (within the pod) in order to pick up the additional log statements?  That might be tricky and you're probably better off just building an EG wheel file. Then building an image that derives from the one you're using except that it installs your wheel and deploy that image.

If you take that approach, you'll probably want to stick to handlers within EG since the others would require build and installation of the notebook (and/or jupyter_client) package(s) and you're then headed down that slippery slope.  Choosing the kernelspecs handler is good since you can easily trigger its invocation.  If you need to look at something other than GET then you'll want to look at enterprise_gateway/services/kernels/handlers.py.

**Reply by averageflamethrowerguy** [reply]
Ah, I totally didn't get that restarting something would be necessary!
Yeah, I'll try building the wheel.

**Reply by averageflamethrowerguy** [reply]
An interesting update here:
I got operational logging in the EG instance (the wheel building worked great, thank you!).
It looks like the IP address headers (X-Real-Ip, X-Forwarded-For) are both being forwarded correctly, all the way to the EG instance. But the ClientIP affinity still isn't taking hold. 
```
Host: <ingress_dns_name>
    X-Request-Id: 3123812a54a59b4b180852194a1714b6
    X-Real-Ip: <my_ip>
    X-Forwarded-For: <my_ip>
    X-Forwarded-Host: <ingress_dns_name>
    X-Forwarded-Port: 443
    X-Forwarded-Proto: https
    X-Scheme: https
    User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:87.0) Gecko/20100101 Firefox/87.0
    ...
```
One possibility is the traffic is being routed independently of the headers. Certainly this would be the case for non-HTTP protocols, so it might make sense if this was the support system for all protocols.
Regardless, this is definitely a big step toward solving the issue because I don't have to worry about headers being changed anymore.

**Comment by averageflamethrowerguy** [comment, marked as answer]
Alright, I have a partial solution to the load balancing issue.
I decided that it would be worthwhile to simplify things and strip out the L7 NGINX load balancer entirely.
I simply exposed the deployment:
```
kubectl expose deployment enterprise-gateway --type=LoadBalancer --name=enterprise-gateway-entry --namespace=enterprise-gateway
```
I then added my SSL cert. 
```
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: MY_SSL_CERT_LINK
```

This resulted in some limited success. Sessions seem to be persisted through the kernel start and kernel retrieval operations. I've been seeing consistent failures on the Restart operations (but those no longer seem to be required in my series of kernel startup operations).
Then provisioning the websocket seems to be done randomly between all available pods. Unfortunately, I was unable to play around with more session stickiness, as I think I need a TCP backend connection to support websockets, and Amazon doesn't allow me to configure sticky sessions on a TCP backend.
I ended up just building some retry logic into my frontend to take care of getting the websocket (my guess is the Jupyter frontend probably does this by default).

Some other possible solutions that I haven't explored (or not deeply) but may be helpful to someone stumbling across this:
1. EG has a external IP configuration (k8sMasterPublicIP) in values.yaml of the helm chart. This sets an external IP address on the NodePort service that exposes the EG deployment by default. Using the NodePort for entry into the cluster may allow you to use `sessionAffinity: ClientIP` to maintain sticky sessions based on ClientIP, By avoiding using a load balancer, you may avoid obfuscating the sourceIP information that I _think_ ClientIP uses for sticky sessions.
I ended up not doing this, because I couldn't figure out how to attach an external IP to the NodePort -- I saw something about needing to expose the node publicly, which I definitely did not want to do. It also seemed like directly attaching an IP was a less robust solution than a load balancer.
2. Use an Application Load Balancer (ALB) instead of a Classic Load Balancer. ALBs supposedly support websocket protocols natively. This might allow using HTTP/Websocket backend protocols rather than raw TCP. It's possible that this opens up more configuration options that allow better session persistence. I'm just even less familiar with the ALB side of things than the CLB side. 

Also, for anyone interested in the Javascript implementation of my retry logic:
```
  // creates a websocket; in case of failure it just times 50 ms and calls itself again (tries up to 20 times).
  createWebsocket(kernelId, iterationNumber) {
    this.dispatch(kernelId, iterationNumber);
    var _this = this

    return new Promise(async (resolve, reject) => {
      var client = new WebSocketClient(`${WEBSOCKET_URL}/api/kernels/${kernelId}/channels`)

      client.onopen = function() {
        _this.actions._updateKernelState('Ready')
        _this.actions._storeWebsocket(client)
        console.log(`Websocket connected to kernel with id ${kernelId}`)
        resolve(client)
      }

      client.onerror = async function() {
        if (iterationNumber < NUMBER_RETRIES) {
          await new Promise((resolve) => setTimeout(resolve, 50))
          try {
            await _this.actions.createWebsocket(kernelId, iterationNumber + 1)
          }
          catch(err) {
            return reject(err)
          }

          return resolve()
        }
        else {
          _this.actions._updateKernelState('Failed')
          reject('Unable to connect to websocket')
        }
      }
    })
  }
```
Note that I'm using [Alt](http://alt.js.org/) to manage actions and state, but the basic business logic should be pretty obvious regardless of framework.
